File: sprint12/JMS/FW/BOJ_11729.cpp
#include <iostream>

using namespace std;

int N;

void Hanoi(int n, int start, int mid, int dest) {
	if (n == 1) {
		cout << start << ' ' << dest << '\n';
		return;
	}

	Hanoi(n - 1, start, dest, mid);
	cout << start << ' ' << dest << '\n';
	Hanoi(n - 1, mid, start, dest);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> N;
	
	cout << (1 << N) - 1 << '\n';
	Hanoi(N, 1, 2, 3);
}

File: sprint12/KMS/FW/BOJ_11729.py
N = int(input())


def hanoi(N, start, middle, dest):
    if N == 1:
        return 1
    result = 0

    result += hanoi(N-1, start, dest, middle)
    result += hanoi(1, start, middle, dest)
    result += hanoi(N-1, middle, start, dest)
    return result


def hanoi_print(N, start, middle, dest):
    if N == 1:
        print(start, dest)
        return 1
    result = 0

    result += hanoi_print(N-1, start, dest, middle)
    result += hanoi_print(1, start, middle, dest)
    result += hanoi_print(N-1, middle, start, dest)
    return result


print(hanoi(N, 1, 2, 3))
hanoi_print(N, 1, 2, 3)
